### DISCUSSION TITLE
How to create a service for all requests, regardless of path?

### DISCUSSION THREAD

**Discussion by hannobraun** [tokio-rs/axum#99] [Q&A] [answered]
Hi everyone,

I'm starting to feel like an idiot, because this seems like it should be simple: How do I create a service that handles all requests, regardless of the path? My use case is an HTTP server that redirects all incoming requests to HTTPS, preserving the path (and host).

All the examples I found use `route` or `nest`, which don't seem to be suited for this. All my attempts to do this without `route` resulted in services that hyper's `serve` method didn't accept. I can of course do it without axum, using something like hyper's `make_service_fn`, but then I don't get to use axum's extractors, I think.

Thank you,
Hanno

**Comment by davidpdrsn** [comment, marked as answer]
Nesting a route at `"/"` should do the trick:

```rust
let app = nest("/", route("/", get(handler)));
```

However this will capture _all_ requests so handles added above it wont be called (routes are matched bottom to top).

**Reply by davidpdrsn** [reply]
I see why you'd have trouble finding this. It isn't exactly intuitive. We're currently considering wildcard routes like `route("*", ...)` but haven't decided anything yet.

**Reply by hannobraun** [reply]
Thank you, @davidpdrsn! I don't think I would have come up with this solution any time soon :grin:

**Reply by shengsheng** [reply]
@davidpdrsn It's not working on my computer, still return 404. Could you please check my code.

```rust
use axum::http::HeaderMap;
use axum::{handler::get, response::Html, Router};
use std::net::SocketAddr;

#[tokio::main]
async fn main() {
    // build our application with a route
    let app = Router::new()
        .nest("/", Router::new().route("/", get(handler_wildcard)));

    // run it
    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    println!("listening on {}", addr);
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
}

async fn handler_wildcard(all_headers: HeaderMap) -> Html<String> {
    Html(format!("{:?}\n", all_headers))
}
```

**Reply by davidpdrsn** [reply]
I don't remember why we needed the additional `route`. Perhaps something changed between 0.1 and 0.2.

This works:

```rust
Router::new().nest("/", get(handler_wildcard))
```

You can also use `Shared` from tower to avoid using a `Router`:

```rust
use axum::http::HeaderMap;
use axum::{handler::Handler, response::Html};
use std::net::SocketAddr;

#[tokio::main]
async fn main() {
    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    println!("listening on {}", addr);
    axum::Server::bind(&addr)
        .serve(tower::make::Shared::new(handler_wildcard.into_service()))
        .await
        .unwrap();
}

async fn handler_wildcard(all_headers: HeaderMap) -> Html<String> {
    Html(format!("{:?}\n", all_headers))
}
```
